File: src/server/policy.js
'use strict';
/**
 * File sharing policies: who may read, write or share which file.
 *
 * Stored as one small JSON file so anyone can open it and read the rules:
 *
 *     {
 *       "report.pdf": {
 *         "owner":   "alice",
 *         "readers": ["bob"],
 *         "writers": [],
 *         "public":  false
 *       }
 *     }
 *
 * Rules, in plain English:
 *   - whoever uploads a file first becomes its owner;
 *   - the owner may always read, overwrite and share it;
 *   - a reader may download it, a writer may overwrite it;
 *   - a public file may be downloaded by anyone;
 *   - a file with no entry yet is free for anyone to claim.
 */

const fsp = require('fs/promises');
const path = require('path');

const config = require('../common/config');
const { bus, POLICY_CHANGED } = require('../common/events');

const READ = 'read';
const WRITE = 'write';

const LOCK_RETRY_MS = 10;
const LOCK_TIMEOUT_MS = 5000;

const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

class PolicyStore {
  constructor(file = config.POLICY_FILE) {
    this.path = file;
    this.lockPath = file + '.lock';
    this._rules = {};
    this._queue = Promise.resolve(); // in-process serialisation
  }

  // -------------------------------------------------------------- persistence
  async _load() {
    try {
      this._rules = JSON.parse(await fsp.readFile(this.path, 'utf8'));
    } catch {
      this._rules = {}; // missing or corrupt: start empty
    }
  }

  async _save() {
    await fsp.mkdir(path.dirname(this.path), { recursive: true });
    await fsp.writeFile(this.path, JSON.stringify(this._rules, null, 2), 'utf8');
  }

  /** Re-read the rules from disk (another node may have changed them). */
  async reload() {
    await this._load();
  }

  /**
   * Guard a read-modify-write of the shared policy file.
   *
   * Two different problems have to be solved at once:
   *
   *  1. Three node PROCESSES all edit the same policies.json. Without a guard,
   *     two nodes finishing an upload at the same moment would each save their
   *     own copy of the whole file and one rule would silently vanish. A lock
   *     file opened with the 'wx' flag - which only one process can create -
   *     serialises them.
   *
   *  2. Inside ONE process, `await` lets two connections interleave halfway
   *     through the read-modify-write. Being single-threaded does not save us
   *     here. The promise queue makes them line up.
   */
  async _exclusive(mutate) {
    const run = this._queue.then(async () => {
      const handle = await this._acquireLockFile();
      try {
        await this._load(); // pick up anything another node just wrote
        const result = await mutate();
        await this._save();
        return result;
      } finally {
        await handle.close();
        await fsp.rm(this.lockPath, { force: true });
      }
    });

    this._queue = run.catch(() => {}); // one failure must not block the rest
    return run;
  }

  async _acquireLockFile() {
    const deadline = Date.now() + LOCK_TIMEOUT_MS;
    for (;;) {
      try {
        return await fsp.open(this.lockPath, 'wx'); // fails if it already exists
      } catch (err) {
        if (err.code !== 'EEXIST') throw err;
        if (Date.now() > deadline) {
          await fsp.rm(this.lockPath, { force: true }); // a crashed node left it
          continue;
        }
        await sleep(LOCK_RETRY_MS);
      }
    }
  }

  allRules() {
    return JSON.parse(JSON.stringify(this._rules)); // plain copy
  }

  // ------------------------------------------------------------------ queries
  ownerOf(filename) {
    const rule = this._rules[filename];
    return rule ? rule.owner : null;
  }

  canRead(user, filename) {
    const rule = this._rules[filename];
    if (!rule) return false; // cannot read a file that was never uploaded
    return Boolean(rule.public) || user === rule.owner
      || (rule.readers || []).includes(user);
  }

  canWrite(user, filename) {
    const rule = this._rules[filename];
    if (!rule) return true; // brand new file: the first uploader claims it
    return user === rule.owner || (rule.writers || []).includes(user);
  }

  /** The set of file names this user is allowed to see. */
  visibleTo(user) {
    return Object.entries(this._rules)
      .filter(([, rule]) => rule.public || user === rule.owner
        || (rule.readers || []).includes(user)
        || (rule.writers || []).includes(user))
      .map(([name]) => name);
  }

  // ------------------------------------------------------------------- writes
  /** Called after a successful upload: create the rule if it is new. */
  async claim(user, filename) {
    const created = await this._exclusive(() => {
      if (this._rules[filename]) return false;
      this._rules[filename] = { owner: user, readers: [], writers: [], public: false };
      return true;
    });

    if (created) bus.publish(POLICY_CHANGED, { file: filename, user, action: 'owner' });
  }

  /** Owner grants read or write access to another user. */
  async share(owner, filename, targetUser, permission) {
    const outcome = await this._exclusive(() => {
      const rule = this._rules[filename];
      if (!rule) return { ok: false, message: 'file not found' };
      if (rule.owner !== owner) {
        return { ok: false, message: 'only the owner can share this file' };
      }

      let bucket;
      if (permission === READ) bucket = (rule.readers ||= []);
      else if (permission === WRITE) bucket = (rule.writers ||= []);
      else return { ok: false, message: 'permission must be read or write' };

      if (!bucket.includes(targetUser)) bucket.push(targetUser);
      return {
        ok: true,
        message: `granted ${permission} access on ${filename} to ${targetUser}`,
      };
    });

    if (outcome.ok) {
      bus.publish(POLICY_CHANGED, {
        file: filename, user: owner, action: 'share',
        target: targetUser, permission,
      });
    }
    return outcome;
  }

  async setPublic(owner, filename, isPublic = true) {
    const outcome = await this._exclusive(() => {
      const rule = this._rules[filename];
      if (!rule) return { ok: false, message: 'file not found' };
      if (rule.owner !== owner) {
        return { ok: false, message: 'only the owner can change visibility' };
      }
      rule.public = Boolean(isPublic);
      return {
        ok: true,
        message: `${filename} is now ${isPublic ? 'public' : 'private'}`,
      };
    });

    if (outcome.ok) {
      bus.publish(POLICY_CHANGED, { file: filename, user: owner, action: 'public' });
    }
    return outcome;
  }
}

module.exports = { PolicyStore, READ, WRITE };

File: src/server/policy.py
"""File sharing policies: who may read, write or share which file.

Stored as one small JSON file so anyone can open it and read the rules:

    {
      "report.pdf": {
        "owner":   "alice",
        "readers": ["bob"],
        "writers": [],
        "public":  false
      }
    }

Rules, in plain English:
  * whoever uploads a file first becomes its owner;
  * the owner may always read, overwrite and share it;
  * a reader may download it, a writer may overwrite it;
  * a public file may be downloaded by anyone;
  * a file with no entry yet is free for anyone to claim.
"""

import contextlib
import json
import os
import threading
import time

from ..common import config
from ..common.events import bus, POLICY_CHANGED

READ = "read"
WRITE = "write"


class PolicyStore:
    def __init__(self, path=None):
        self.path = path or config.POLICY_FILE
        self._lock = threading.Lock()
        self._rules = {}
        self._load()

    # ------------------------------------------------------------ persistence
    def _load(self):
        if os.path.exists(self.path):
            try:
                with open(self.path, "r", encoding="utf-8") as fh:
                    self._rules = json.load(fh)
            except (ValueError, OSError):
                self._rules = {}

    def _save(self):
        os.makedirs(os.path.dirname(self.path), exist_ok=True)
        with open(self.path, "w", encoding="utf-8") as fh:
            json.dump(self._rules, fh, indent=2)

    @contextlib.contextmanager
    def _exclusive(self):
        """Guard a read-modify-write of the shared policy file.

        Every node runs in its own process but they all edit the same
        policies.json. Without this, two nodes finishing an upload at the same
        moment would each save their own copy and one rule would be lost. A
        simple lock file (created with O_EXCL, which only one process can win)
        serialises them.
        """
        lock_path = self.path + ".lock"
        deadline = time.time() + 5.0
        while True:
            try:
                fd = os.open(lock_path, os.O_CREAT | os.O_EXCL | os.O_RDWR)
                break
            except FileExistsError:
                if time.time() > deadline:
                    os.remove(lock_path)   # a crashed node left it behind
                    continue
                time.sleep(0.01)
        try:
            self._load()      # pick up anything another node just wrote
            yield
            self._save()
        finally:
            os.close(fd)
            try:
                os.remove(lock_path)
            except OSError:
                pass

    def reload(self):
        """Re-read the rules from disk (other nodes may have changed them)."""
        with self._lock:
            self._load()

    def all_rules(self):
        with self._lock:
            return json.loads(json.dumps(self._rules))  # plain copy

    # ---------------------------------------------------------------- queries
    def owner_of(self, filename):
        with self._lock:
            rule = self._rules.get(filename)
            return rule["owner"] if rule else None

    def can_read(self, user, filename):
        with self._lock:
            rule = self._rules.get(filename)
            if rule is None:
                return False  # cannot read a file that was never uploaded
            return (rule.get("public", False)
                    or user == rule["owner"]
                    or user in rule.get("readers", []))

    def can_write(self, user, filename):
        with self._lock:
            rule = self._rules.get(filename)
            if rule is None:
                return True  # brand new file: the first uploader claims it
            return user == rule["owner"] or user in rule.get("writers", [])

    # ----------------------------------------------------------------- writes
    def claim(self, user, filename):
        """Called after a successful upload: create the rule if it is new."""
        with self._lock, self._exclusive():
            created = filename not in self._rules
            if created:
                self._rules[filename] = {
                    "owner": user,
                    "readers": [],
                    "writers": [],
                    "public": False,
                }
        if created:
            bus.publish(POLICY_CHANGED, file=filename, user=user, action="owner")

    def share(self, owner, filename, target_user, permission):
        """Owner grants read or write access to another user."""
        with self._lock, self._exclusive():
            rule = self._rules.get(filename)
            if rule is None:
                return False, "file not found"
            if rule["owner"] != owner:
                return False, "only the owner can share this file"

            if permission == READ:
                bucket = rule.setdefault("readers", [])
            elif permission == WRITE:
                bucket = rule.setdefault("writers", [])
            else:
                return False, "permission must be read or write"

            if target_user not in bucket:
                bucket.append(target_user)

        bus.publish(POLICY_CHANGED, file=filename, user=owner,
                    action="share", target=target_user, permission=permission)
        return True, "granted %s access on %s to %s" % (permission, filename, target_user)

    def set_public(self, owner, filename, public=True):
        with self._lock, self._exclusive():
            rule = self._rules.get(filename)
            if rule is None:
                return False, "file not found"
            if rule["owner"] != owner:
                return False, "only the owner can change visibility"
            rule["public"] = bool(public)
        bus.publish(POLICY_CHANGED, file=filename, user=owner, action="public")
        return True, "%s is now %s" % (filename, "public" if public else "private")

    def visible_to(self, user):
        """The set of file names this user is allowed to see."""
        with self._lock:
            return [name for name, rule in self._rules.items()
                    if rule.get("public") or user == rule["owner"]
                    or user in rule.get("readers", [])
                    or user in rule.get("writers", [])]
